timemachine/cpp/src/kernels/k_exchange.cuh

```diff
@@ -78,4 +108,83 @@ void __global__ k_store_accepted_log_probability(
     }
 }
 
+template <typename RealType, bool Negated>
+void __global__ k_adjust_weights(
+    const int N,
+    const int num_target_mols,
+    const int mol_size,
+    const int *__restrict__ mol_atoms_idxs,
+    const int *__restrict__ mol_offsets,
+    const RealType *__restrict__ per_atom_energies,
+    const RealType inv_kT, // 1 / kT
+    RealType *__restrict__ log_weights) {
+    int mol_idx = blockIdx.x * blockDim.x + threadIdx.x;
+    while (mol_idx < num_target_mols) {
+
+        RealType current_log_weight = log_weights[mol_idx];
+        __int128 weight_accumulator = 0;
+
+        int mol_start = mol_offsets[mol_idx];
+        int mol_end = mol_offsets[mol_idx + 1];
+        int min_atom_idx = mol_atoms_idxs[mol_start];
+        int max_atom_idx = mol_atoms_idxs[mol_end - 1];
+
+        // A loop that in the case of water will be 3x3
+        for (int i = 0; i < mol_size; i++) {
+            for (int j = min_atom_idx; j <= max_atom_idx; j++) {
+                weight_accumulator += FLOAT_TO_FIXED_ENERGY<RealType>(inv_kT * per_atom_energies[i * N + j]);
+            }
+        }
+
+        weight_accumulator = Negated ? FLOAT_TO_FIXED_ENERGY<RealType>(current_log_weight) - weight_accumulator
+                                     : FLOAT_TO_FIXED_ENERGY<RealType>(current_log_weight) + weight_accumulator;
+
+        log_weights[mol_idx] =
+            fixed_point_overflow(weight_accumulator) ? INFINITY : FIXED_ENERGY_TO_FLOAT<RealType>(weight_accumulator);
+
+        mol_idx += gridDim.x * blockDim.x;
+    }
+}
+
+template <typename RealType, int THREADS_PER_BLOCK>
+void __global__ k_set_sampled_weight(
+    const int N,
+    const int mol_size,
+    const int num_samples,
+    const int *__restrict__ samples, // [num_samples]
+    const int *__restrict__ target_atoms,
+    const int *__restrict__ mol_offsets,
+    const RealType *__restrict__ per_atom_energies,
+    const RealType inv_kT, // 1 / kT
+    RealType *__restrict__ log_weights) {
+    static_assert(THREADS_PER_BLOCK <= 512 && (THREADS_PER_BLOCK & (THREADS_PER_BLOCK - 1)) == 0);
```

Comment:
What's the purpose of the second part `(THREADS_PER_BLOCK & (THREADS_PER_BLOCK - 1)) == 0`? Check that it's a power of 2?

Reply:
Correct, as this kernel expects everything is a power of 2